Src/ILGPU/Runtime/ExchangeBuffer.cs

```diff
@@ -381,4 +445,122 @@ protected override void Dispose(bool disposing)
+    /// </summary>
+    /// <typeparam name="T">The element type.</typeparam>
+    /// <remarks>Members of this class are not thread safe.</remarks>
+    public sealed unsafe class ExchangeBuffer2D<T> : ExchangeBufferBase<T, Index2>
+        where T : unmanaged
+    {
+        #region Instance
+
+        /// <summary>
+        /// Initializes this memory buffer.
+        /// </summary>
+        /// <param name="buffer">The underlying memory buffer.</param>
+        /// <param name="mode">The current buffer allocation mode.</param>
+        internal ExchangeBuffer2D(MemoryBuffer<T, Index2> buffer, ExchangeBufferMode mode)
+            : base(buffer, mode)
+        {
+            var baseView = new ArrayView<T>(cpuMemory, 0, buffer.Length);
+            CPUView = new ArrayView<T, Index2>(baseView, buffer.Extent);
+
+            // Cache local data
+            Buffer = buffer;
+            NativePtr = buffer.NativePtr;
+            View = buffer.View;
+            Extent = buffer.Extent;
+        }
+
+        #endregion
+
+        #region Methods
+
+        /// <summary>
+        /// Gets this buffer as a 2D array from the accelerator
+        /// </summary>
+        /// <returns>The array containing all the elemnts in the buffer</returns>
+        public T[,] GetAs2DArray()
+        {
+            var buffer = new MemoryBuffer2D<T>(Buffer);
+            var array = buffer.GetAs2DArray();
+            buffer.Dispose();
+            return array;
+        }
+
+        #endregion
+    }
+
+    /// <summary>
+    /// 3D implementation of <see cref="ExchangeBufferBase{T, TIndex}"/>
+    /// </summary>
+    /// <typeparam name="T">The element type.</typeparam>
+    /// <remarks>Members of this class are not thread safe.</remarks>
+    public sealed unsafe class ExchangeBuffer3D<T> : ExchangeBufferBase<T, Index3>
+        where T : unmanaged
+    {
+        #region Instance
+
+        /// <summary>
+        /// Initializes this memory buffer.
+        /// </summary>
+        /// <param name="buffer">The underlying memory buffer.</param>
+        /// <param name="mode">The current buffer allocation mode.</param>
+        internal ExchangeBuffer3D(MemoryBuffer<T, Index3> buffer, ExchangeBufferMode mode)
+            : base(buffer, mode)
+        {
+            var baseView = new ArrayView<T>(cpuMemory, 0, buffer.Length);
+            CPUView = new ArrayView<T, Index3>(baseView, buffer.Extent);
+
+            // Cache local data
+            Buffer = buffer;
+            NativePtr = buffer.NativePtr;
+            View = buffer.View;
+            Extent = buffer.Extent;
+        }
+
+        #endregion
+
+        #region Methods
+
+        /// <summary>
+        /// Gets this buffer as a 3D array from the accelerator
+        /// </summary>
+        /// <returns>The array containing all the elemnts in the buffer</returns>
+        public T[,,] GetAs3DArray() => new MemoryBuffer3D<T>(Buffer).GetAs3DArray();
```

Comment:
Please be very careful about the allocation of temporary `MemoryBufferYD<T>` instances. Once your new `MemoryBuffer3D<T>` instance will be garbage collected, it will invoke its finalizer and dispose the associated `Buffer` instance. In this case, your buffer will be freed although there is still a reference in the scope of the `ExchangeBuffer` object.

Reply:
Ok, I will implement it how I did with the 2D exchange Buffer then, I free it there.